src/js/components/SidebarLabelsFilter.js

```diff
@@ -0,0 +1,217 @@
+            return label.key === key && label.value === value;
+          });
+
+          if (0 > index) {
+            memo = memo.concat([{key, value}]);
+          }
+        });
+      }
+      if (item instanceof ServiceTree) {
+        memo = memo.concat(item.getLabels());
+      }
+
+      return memo;
+    },  [])
+    .sort((a, b) => a.key.localeCompare(b.key));
+  }
+
+  getLabelsDropdown() {
+    let {state} = this;
+    let availableLabels = state.availableLabels.map(function (label, i) {
+      let labelText = `${label.key} : ${label.value}`;
+
+      return Object.assign({}, label, {
+        id: `filter-label-${i}`,
+        html: (
+          <a className="text-overflow" title={labelText}>
+            {labelText}
+          </a>
+        )
+      });
+    });
+
+    let labelOptions = [{
+      className: 'hidden',
+      id: '0',
+      html: 'Labels',
+      selectable: false
+    }].concat(availableLabels);
+
+    return (
+      <Dropdown
+        buttonClassName="button button-inverse dropdown-toggle button-wide button-split-content"
+        dropdownMenuClassName="dropdown-menu inverse"
+        dropdownMenuListClassName="dropdown-menu-list"
+        dropdownMenuListItemClassName="clickable"
+        items={labelOptions}
+        onItemSelection={this.handleActionSelection}
+        persistentID="0"
+        transition={true}
+        transitionName="dropdown-menu"
+        wrapperClassName="dropdown dropdown-wide" />
+    );
+  }
+
+  getSelectedLabels() {
+    let {selectedLabels} = this.state;
+
+    if (selectedLabels == null || selectedLabels.length === 0) {
+      return null;
+    }
+
+    const labelClassName = classNames(
+      'label-pill badge text-muted text-overflow inverse'
+    );
+
+    const removeLabelClassNames = classNames(
+      'badge text-muted inverse clickable text-align-center remove-filter'
+    );
+
+    let labelNodes = selectedLabels.map(({key, value}, i) => {
+      let labelText = `${key} : ${value}`;
+      return (
+        <li className={labelClassName} key={i} title={labelText}>
+          <div className="text-overflow pill-wrap">
+            <span className="text-overflow">{labelText}</span>
+            <a className={removeLabelClassNames}
+               onClick={this.handleActionSelection.bind(this, {key, value})}>
+              x
+            </a>
+          </div>
+        </li>
+      );
+    });
+
+    if (labelNodes.length === 0) {
+      return null;
+    }
+
+    const labelsListClassNames = classNames(
+      'container-pod container-pod-super-super-short flush-bottom list-unstyled'
+    );
+
+    return (
+      <ul className={labelsListClassNames}>
+        {labelNodes}
+      </ul>
+    )
+  }
+
+  handleActionSelection({key, value}) {
+    const {selectedLabels} = this.state;
+    const stringify = JSON.stringify;
+    let nextSelectedLabels = selectedLabels.slice();
+
+    let labelIndex = selectedLabels.findIndex(function (item) {
```

Comment:
Ah, the `labelIndex` could be a `const` too. :)
But I don't care, only change if you like.